composer/bin/mago

```diff
@@ -0,0 +1,137 @@
+                            break;
+                        case 'arm':
+                        case 'armv7':
+                            if (strpos(file_get_contents('/proc/cpuinfo'), 'hard') !== false) {
+                                $os_suffix = 'musleabihf';
+                            } else {
+                                $os_suffix = 'musleabi';
+                            }
+                            break;
+                        default:
+                            throw new \RuntimeException("Unsupported architecture for musl: {$arch_name}");
+                    }
+                } else {
+                    switch ($arch) {
+                        case 'x86_64':
+                        case 'aarch64':
+                        case 'i686':
+                        case 'powerpc':
+                        case 'powerpc64':
+                        case 'powerpc64le':
+                        case 's390x':
+                            $os_suffix = 'gnu';
+                            break;
+                        case 'arm':
+                        case 'armv7':
+                            if (strpos(file_get_contents('/proc/cpuinfo'), 'hard') !== false) {
+                                $os_suffix = 'gnueabihf';
+                            } else {
+                                $os_suffix = 'gnueabi';
+                            }
+                            break;
+                        default:
+                            throw new \RuntimeException("Unsupported architecture for glibc: {$arch_name}");
+                    }
+                }
+            } else {
+                $os_suffix = 'musl';
+            }
+            break;
+        case 'freebsd':
+            break;
+        default:
+            throw new \RuntimeException("Unsupported operating system: {$os}. Please open an issue on GitHub.");
+    }
+
+    $target_triple = $os_suffix ? "{$arch}-{$vendor}-{$os}-{$os_suffix}" : "{$arch}-{$vendor}-{$os}";
+    $executable = __DIR__ . "/{$version}/mago-{$version}-{$target_triple}/mago{$executable_extension}";
+
+    if (!file_exists($executable)) {
+        throw new \RuntimeException("Unable to find mago executable at {$executable}. Please run 'composer mago:install-assets'");
+    }
+
+    $args = $_SERVER['argv'];
+    array_shift($args);
+
+    $command = escapeshellcmd($executable);
+    if ($args) {
+        $command .= ' ' . implode(' ', array_map(escapeshellarg(...), $args));
+    }
+
+    // Run command
+    $pipes = [];
+    $process = @proc_open(
+        $command,
+        // Map process to current io
+        $descriptorspec = array(
+            0 => array('file', 'php://stdin', 'r'),
+            1 => array('file', 'php://stdout', 'w'),
+            2 => array('file', 'php://stderr', 'w'),
+        ),
+        $pipes
+    );
+
+    if (!is_resource($process)) {
+        echo('Unable to start mago process');
+        exit(1);
+    }
+
```

Comment:
can we have the php script move the real executable to its place or symlinking it, which will avoid the overhead on every use?

Reply:
Yes and no...

That would make it not possible to execute on both the host system and e.g. inside docker at the same time which it can do right now.

I can of course move the logic the the composer command and even only download the binaries for that specific platform, but it will crash when its being used like that on multiple architectures.

I've thought about it as well but I'm not sure how we want it to be used in that kind of situations.

Reply:
> will crash when its being used like that on multiple architectures.

why? can't we have it execute on every install/update and download the binary then? 🤔

Reply:
It will then only download the binary for the architecture you are currently run composer on.

If you are composer installing on e.g. Mac it will have the Darwin binary. But if you also work with docker on the project, you might need some Linux version as well for the same project: if you want to run the mago command from both host and guest. So in that situation where you composer installed locally, and execute Mago grom the link guest container, it will run the Darwin version on linux and crash. People are pretty used to having the php executables working cross system.

Since this pr also introduces a composer command for installing the assets, it's pretty easy to switch versions that way though. So maybe going for one version might be OK as well.